### PR TITLE
Basic Auth for API v3

### PR DESCRIPTION
**Pull request by machisuji**
WP [#19294](https://community.openproject.org/work_packages/19294)

### CODE REVIEW

**Comment by TeatroIO** [comment]
I've prepared a stage to preview changes. <a href='http://machisuji-feature-basic-auth.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com' target='_blank'>Open stage</a> or <a href='http://machisuji-feature-basic-auth.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com/teatro.log' target='_blank'>view logs</a>.

**Comment by floriank** [comment]
changed the title as the label "work in progress" designates work in progress already.

File: Gemfile [hunk 1]
```diff
@@ -38,6 +38,9 @@ gem 'omniauth'
 gem 'request_store', "~> 1.1.0"
 gem 'gravatar_image_tag', '~> 1.2.0'
 
+gem 'warden', '~> 1.2'
```
Comment (linki) [inline Gemfile:41]: let's change that to `~> 1.2.3` so we follow the semantic versioning paradigm more closely

Comment (myabc) [inline Gemfile:41]: @linki take a look at our discussion over here: https://github.com/opf/openproject/pull/2104#issuecomment-99760061

File: app/models/system_user.rb [hunk 2]
```diff
@@ -62,7 +62,7 @@ def available_custom_fields
 
   # There should be only one SystemUser in the database
   def validate_unique_system_user
-    errors.add :base, 'A SystemUser already exists.' if SystemUser.find(:first)
+    errors.add :base, 'A SystemUser already exists.' if SystemUser.first
```
Comment (linki) [inline app/models/system_user.rb]: then let's use `any?` :)

```
[5] pry(main)> Augur::Token.first
  Augur::Token Load (2.2ms)  SELECT  "tokens".* FROM "tokens"   ORDER BY "tokens"."id" ASC LIMIT 1
=> nil
[6] pry(main)> Augur::Token.any?
   (0.4ms)  SELECT COUNT(*) FROM "tokens"
=> false
[8] pry(main)> Augur::Token.create! type: 'forgotten'
=> #<Augur::Token id: 1, code: nil, type: "forgotten", account_id: nil, created_at: "2015-05-07 13:57:39", updated_at: "2015-05-07 13:57:39">
[9] pry(main)> Augur::Token.any?
   (0.7ms)  SELECT COUNT(*) FROM "tokens"
=> true
```

tested with active record 4

File: config/initializers/warden.rb [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+  [:session,            Warden::Strategies::Session]
+]
+
+strategies.each do |name, clazz|
+  Warden::Strategies.add name, clazz
+end
+
+include OpenProject::Authentication::Scope
+
+OpenProject::Authentication.update_strategies(API_V3) do |_strategies|
```
Comment (oliverguenther) [inline config/initializers/warden.rb:22]: Using an included const like that is something I've never seen before.
It does not prevent passing an invalid / unknown scope to `update_strategies` (with a symbol) and using :api_v3 instead would spare you the include.

Is that a common style I've missed elsewhere or does it serve another purpose I cannot see?

Comment (linki) [inline config/initializers/warden.rb:22]: why no symbol for `:api_v3`?

File: lib/api/root.rb [hunk 4]
```diff
@@ -59,19 +63,32 @@ def call(object, _env)
...
+
+        if Setting.login_required? && not_logged_in?
           raise API::Errors::Unauthenticated
         end
       end
 
+      def not_logged_in?
+        # An admin SystemUser is anonymous but still a valid user to be logged in.
+        current_user.nil? || (!current_user.admin? && current_user.anonymous?)
+      end
```
Comment (linki) [inline lib/api/root.rb:90]: thx for extracting and giving it a name :+1:

File: config/initializers/warden.rb (hunk 3 above)
Comment (machisuji) [inline config/initializers/warden.rb:22]: It's probably not common. And there's nothing wrong with just passing `:api_v3` directly, I suppose.
I've just used the constant as this way I already notice if I used an invalid scope when the file is required due to a name error which means the software doesn't even start.

File: lib/open_project/authentication.rb [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+      #
+      # @yield [strategies] A block returning the strategies to be used for this scope.
+      # @yieldparam [Array] The strategies currently used by this scope. May be empty.
+      # @yieldreturn [Array] The strategies to be used by this scope.
+      def update_strategies(scope, store: nil, &block)
+        raise ArgumentError, "invalid scope: #{scope}" unless Scope.values.include? scope
+
+        current_strategies = Array(Manager.scope_strategies[scope])
+
+        Manager.store_defaults[scope] = store unless store.nil?
```
Comment (linki) [inline lib/open_project/authentication.rb:25]: `if store` :astonished:

Comment (machisuji) [inline lib/open_project/authentication.rb:25]: The `.nil?` is intentional. If I use `if store` I can't set it to false.

**Comment by linki** [comment]
i would bundle the warden strategies under `lib/open_project/authentication` so auth related stuff sits together. e.g. in `lib/open_project/authentication/strategies` or `lib/open_project/authentication/warden/strategies` or `lib/open_project/authentication/strategies/warden`

File: lib/warden/strategies/user_basic_auth.rb [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+    ##
+    # Allows users to authenticate using their API key via basic auth.
+    # Note that in order for a user to be able to generate one
+    # `Setting.rest_api_enabled` has to be `1`.
+    #
+    # The basic auth credentials are expected to contain the literal 'apikey'
+    # as the user name and the API key as the password.
+    class UserBasicAuth < BasicAuth
+      def self.user
+        'apikey'
```
Comment (oliverguenther) [inline lib/warden/strategies/user_basic_auth.rb]: As `GlobalBasicAuth` takes precedence  and is tried first, then (if I'm not mistaken) `UserBasicAuth` will break if someone is ignorant enough to use _apikey_ as the global username.
You could check the global username and prevent that value from being taken. If so, this line should be extracted somewhere (possibly to the Settings).

Comment (machisuji) [inline lib/warden/strategies/user_basic_auth.rb]: Good point.

File: lib/open_project/authentication.rb (hunk 5 above)
Comment (linki) [inline lib/open_project/authentication.rb:25]: is there a difference between setting it to `nil` or `false`? if not you could default it to `false` and always set it here

Comment (machisuji) [inline lib/open_project/authentication.rb:25]: The point is that one might now want to update the store setting and just leave it as it is. Potentially while not even knowing what the current value is.

File: config/initializers/warden.rb (hunk 3 above)
Comment (oliverguenther) [inline config/initializers/warden.rb:22]: Okay, I get that and agree with you (even though it doesn't prevent someone else to use a symbol when another strategy is introduced and bypassing the early error).

**Comment by oliverguenther** [comment]
I agree with @linki that warden et al. would best reside under `open_project/authentication`, joined later on by omniauth code - if your goal is to unify all authentication schemes with omniauth and warden (which I strongly support and anticipate :grinning: )

Except the noted minor remarks, this looks very good overall and is a great step towards unification.

**Comment by machisuji** [comment]
Yes, unification is the plan. I agree. Will move the code there. Thanks for having a look at it!

File: Gemfile (hunk 1 above)
Comment (linki) [inline Gemfile:41]: didn't we decide https://github.com/opf/openproject/pull/2104#issuecomment-99890764? i think `warden` is pretty stable.

**Comment by machisuji** [comment]
Now that I moved the strategies the fully qualified names are wee bit long, but oh well.

**Comment by machisuji** [comment]
Re-opening against dev.

**Pull request closed without merging**